### PR TITLE
[fix] support different geometry types

### LINKED ISSUE #1139: Compilation bug in >= 1.76 from detail::overlay::collinear::handle_as_equal
I experience some problems when adapting [Lanelet2 library](https://github.com/fzi-forschungszentrum-informatik/Lanelet2) to boost v1.78.0 , one of which is caused by a bug introduced with release v1.76.0 and has not been fixed yet.

### When was this introduced
Introduced with 84d3a8d9ceb49489e8c230b92843326f5907417b, specifically this line: https://github.com/boostorg/geometry/commit/84d3a8d9ceb49489e8c230b92843326f5907417b#diff-21d011ddccb292362e430e14bebe3673bfa5e10f432ee3a85ede067fe8f140c6R993 which is still existing in the latest boost version as well.  

### Identified issue
You're using a conditional operator, but `range_q` and `range_p` can be of a different type.

In my case, they were `Eigen::Matrix<double, 2, 1>` and `Eigen::Matrix<double, 2, 1, 2>`, i.e. aligned and non-aligned 2D Eigen-vectors.

### Proposed fix
The fix is quite simple. This would be a proposal for the current version, but applicable to all other releases as well
```diff
-        auto const dm = fun::distance_measure(info.intersections[1],
-                arrival_p == 1 ? range_q.at(1) : range_p.at(1));
+        auto const dm = arrival_p == 1
+               ? fun::distance_measure(info.intersections[1], range_q.at(1))
+               : fun::distance_measure(info.intersections[1], range_p.at(1));
```
It would be nice if this could be backported to all affected revisions as well with a patch-release. I personally would be happy if this was provided "only" from v1.78.0 onward.

### PR DESCRIPTION
**Pull request by barendgehrels**
This fixes #1139 , and more required code changes.

Including unit test.

### CODE REVIEW

File: include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp [hunk 1]
```diff
@@ -1032,8 +1035,9 @@ struct collinear : public base_turn_handler
             return false;
         }
 
-        auto const dm = fun::distance_measure(info.intersections[1],
-                arrival_p == 1 ? range_q.at(1) : range_p.at(1));
+       auto const dm = arrival_p == 1
+              ? fun::distance_measure(info.intersections[1], range_q.at(1))
+              : fun::distance_measure(info.intersections[1], range_p.at(1));
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/overlay/get_turn_info.hpp:1040]: This is the fix proposed in #1139

File: include/boost/geometry/algorithms/detail/sections/sectionalize.hpp [hunk 2]
```diff
@@ -889,26 +889,6 @@ inline void sectionalize(Geometry const& geometry,
...
-        typename geometry::robust_point_type
-        <
-            typename geometry::point_type<Geometry>::type,
-            RobustPolicy
-        >::type
-    >::type ctype2;
-
-    BOOST_STATIC_ASSERT((std::is_same<ctype1, ctype2>::value));
-
-
```
Comment (barendgehrels) [inline include/boost/geometry/algorithms/detail/sections/sectionalize.hpp:911]: This had to go, these types are not always the same.

**Comment by barendgehrels** [comment]
@awulkiew @vissarion OK to merge?

File: include/boost/geometry/algorithms/detail/overlay/append_no_dups_or_spikes.hpp [hunk 3]
```diff
@@ -99,7 +100,16 @@ inline void append_no_dups_or_spikes(Range& range, Point const& point,
         return;
     }
 
-    traits::push_back<Range>::apply(range, point);
+    auto append = [](auto& r, const auto& p)
+    {
+        using point_t = typename boost::range_value<Range>::type;
+        point_t rp;
+        geometry::detail::conversion::convert_point_to_point(p, rp);
+        traits::push_back<Range>::apply(r, rp);
```
Comment (awulkiew) [inline include/boost/geometry/algorithms/detail/overlay/append_no_dups_or_spikes.hpp:108]: The point could be moved here. In practice this wouldn't make any difference but if someone used e.g. Multiprecision coordinates we'd avoid copying.

**Review by awulkiew — APPROVED** [review]
Yes, LGTM, thanks!

**Comment by vissarion** [comment]
Sorry for the delays, I can review it tomorrow.

File: include/boost/geometry/algorithms/detail/direction_code.hpp [hunk 4]
```diff
@@ -87,42 +87,59 @@ struct direction_code_impl<cartesian_tag>
...
+                "Not implemented for different units.",
+                units_sa_t, units_p_t);
+        }
+
+        using coor_sa_t = typename coordinate_type<PointSegmentA>::type;
+        using coor_sb_t = typename coordinate_type<PointSegmentB>::type;
+        using coor_p_t = typename coordinate_type<Point2>::type;
+
+        // Declare unit type (equal for all types) and calc type (coerced to most precise)
+        using units_t = typename cs_angular_units<Point2>::type;
```
Comment (vissarion) [inline include/boost/geometry/algorithms/detail/direction_code.hpp:113]: This is `units_p_t` above. Should we reuse that type?

File: include/boost/geometry/algorithms/detail/direction_code.hpp [hunk 5]
```diff
@@ -87,42 +87,59 @@ struct direction_code_impl<cartesian_tag>
...
+                (std::is_same<units_sa_t, units_p_t>::value),
+                "Not implemented for different units.",
+                units_sa_t, units_p_t);
+        }
+
+        using coor_sa_t = typename coordinate_type<PointSegmentA>::type;
+        using coor_sb_t = typename coordinate_type<PointSegmentB>::type;
+        using coor_p_t = typename coordinate_type<Point2>::type;
+
+        // Declare unit type (equal for all types) and calc type (coerced to most precise)
```
Comment (vissarion) [inline include/boost/geometry/algorithms/detail/direction_code.hpp:112]: What happens if unit type is not the same for all types? Should we assert here?

File: include/boost/geometry/algorithms/detail/direction_code.hpp [hunk 6]
```diff
@@ -140,12 +157,12 @@ struct direction_code_impl<spherical_equatorial_tag>
...
-        calc_t mx = is_a_pole || is_b_pole || is_p_pole ?
-                    c0 :
-                    (std::min)(is_antilon1 ? c0 : math::abs(dlon1),
-                               is_antilon2 ? c0 : math::abs(dlon2));
-        calc_t my = (std::min)(math::abs(dlat1),
-                               math::abs(dlat2));
+        calc_t const mx = is_a_pole || is_b_pole || is_p_pole ?
+                            c0 :
+                            (std::min)(is_antilon1 ? c0 : math::abs(dlon1),
+                                    is_antilon2 ? c0 : math::abs(dlon2));
```
Comment (vissarion) [inline include/boost/geometry/algorithms/detail/direction_code.hpp:163]: should this be aligned with `is_antilon1` in above line?

File: include/boost/geometry/algorithms/detail/direction_code.hpp [hunk 7]
```diff
@@ -188,27 +205,27 @@ struct direction_code_impl<spherical_equatorial_tag>
 template <>
 struct direction_code_impl<spherical_polar_tag>
 {
-    template <typename Point1, typename Point2>
-    static inline int apply(Point1 segment_a, Point1 segment_b,
+    template <typename PointSegmentA, typename PointSegmentB, typename Point2>
+    static inline int apply(PointSegmentA segment_a, PointSegmentB segment_b,
                             Point2 p)
     {
         typedef math::detail::constants_on_spheroid
```
Comment (vissarion) [inline include/boost/geometry/algorithms/detail/direction_code.hpp:212]: Since you replacing typedefs maybe also replace those?

**Review by vissarion — APPROVED** [review]
Thanks, I am OK with merging, with few minor comments.

File: include/boost/geometry/algorithms/detail/overlay/append_no_dups_or_spikes.hpp (hunk 3 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/overlay/append_no_dups_or_spikes.hpp:108]: :heavy_check_mark:

File: include/boost/geometry/algorithms/detail/direction_code.hpp (hunk 4 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/direction_code.hpp:113]: :heavy_check_mark:

File: include/boost/geometry/algorithms/detail/direction_code.hpp (hunk 5 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/direction_code.hpp:112]: The static assertion are above (line 98/102)

File: include/boost/geometry/algorithms/detail/direction_code.hpp (hunk 6 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/direction_code.hpp:163]: Indeed. :heavy_check_mark: 
Updated (I usually align in another way, changed to that)

File: include/boost/geometry/algorithms/detail/direction_code.hpp (hunk 7 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/direction_code.hpp:212]: :heavy_check_mark: thanks

File: include/boost/geometry/algorithms/detail/direction_code.hpp (hunk 4 above)
Comment (barendgehrels) [reply include/boost/geometry/algorithms/detail/direction_code.hpp:113]: Put it back, the one above was in a separate block (specifically for the assertions). I could move that too, but then it's actually clearer like it is now.

**Pull request merged**